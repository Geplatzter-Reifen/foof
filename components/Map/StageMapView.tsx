import { Location, Stage } from "@/database/model/model";
import MapboxGL from "@rnmapbox/maps";
import { StageMapLine } from "@/components/Map/StageMapLine";
import { StyleSheet } from "react-native";
import customStyles from "@/constants/styles";

export default function StageMapView({
  stage,
  locations,
}: {
  stage: Stage;
  locations: Location[];
}) {
  // Find the outermost coordinates & set as bounds
  let minLat = Infinity,
    maxLat = -Infinity,
    minLng = Infinity,
    maxLng = -Infinity;
  locations.forEach((loc) => {
    if (loc.latitude < minLat) minLat = loc.latitude;
    if (loc.latitude > maxLat) maxLat = loc.latitude;
    if (loc.longitude < minLng) minLng = loc.longitude;
    if (loc.longitude > maxLng) maxLng = loc.longitude;
  });
  const bounds = {
    ne: [maxLng, maxLat],
    sw: [minLng, minLat],
  };

  return (
    <MapboxGL.MapView
      scaleBarEnabled={false}
      style={{ ...styles.map, ...customStyles.basicCard }}
    >
      {/* Kamera h√§lt sich an die Bounds */}
      <MapboxGL.Camera
        minZoomLevel={5}
        animationMode="none"
        bounds={bounds}
        padding={{
          paddingTop: 60,
          paddingBottom: 60,
          paddingLeft: 60,
          paddingRight: 60,
        }}
      />
      {/* Linie, die die gefahrene Strecke der Etappe anzeigt */}
      <StageMapLine stage={stage} stageLocations={locations} key={stage.id} />
    </MapboxGL.MapView>
  );
}

const styles = StyleSheet.create({
  map: {
    flex: 1,
  },
});
