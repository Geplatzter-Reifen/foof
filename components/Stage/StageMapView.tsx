import { Location } from "@/database/model/model";
import MapboxGL from "@rnmapbox/maps";
import StageMapLine from "@/components/Tour/StageMapLine";
import { StyleSheet } from "react-native";
import customStyles from "@/constants/styles";

export default function StageMapView({
  stageId,
  locations,
}: {
  stageId: string;
  locations: Location[];
}) {
  // Find the outermost coordinates & set as bounds
  let minLat = Infinity,
    maxLat = -Infinity,
    minLng = Infinity,
    maxLng = -Infinity;
  locations.forEach((loc) => {
    if (loc.latitude < minLat) minLat = loc.latitude;
    if (loc.latitude > maxLat) maxLat = loc.latitude;
    if (loc.longitude < minLng) minLng = loc.longitude;
    if (loc.longitude > maxLng) maxLng = loc.longitude;
  });
  const bounds = {
    ne: [maxLng, maxLat],
    sw: [minLng, minLat],
  };

  return (
    <MapboxGL.MapView
      scaleBarEnabled={false}
      style={{ ...styles.map, ...customStyles.basicCard }}
    >
      {/* Kamera h√§lt sich an die Bounds */}
      <MapboxGL.Camera
        minZoomLevel={5}
        animationMode="none"
        bounds={bounds}
        padding={{
          paddingTop: 60,
          paddingBottom: 60,
          paddingLeft: 60,
          paddingRight: 60,
        }}
      />
      {/* Linie, die die gefahrene Strecke der Etappe anzeigt */}
      {locations.length > 1 && (
        <StageMapLine locations={locations} stageId={stageId} key={stageId} />
      )}
    </MapboxGL.MapView>
  );
}

const styles = StyleSheet.create({
  map: {
    flex: 1,
  },
});
