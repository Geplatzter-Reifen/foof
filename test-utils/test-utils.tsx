// https://testing-library.com/docs/react-testing-library/setup/#add-custom-queries
import React from "react";
import { render, RenderOptions } from "@testing-library/react-native";
import { FontAwesomeIconsPack } from "@/components/Font/fontAwesome";
import { mapping } from "@eva-design/eva";
import {
  IconRegistry,
  ApplicationProvider,
  ModalService,
} from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { foofDarkTheme, foofLightTheme } from "@/constants/custom-theme";
import { library } from "@fortawesome/fontawesome-svg-core";
import { far } from "@fortawesome/free-regular-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { fab } from "@fortawesome/free-brands-svg-icons";

// Iconpacks laden
library.add(far, fas, fab);

// ModalService wird verwendet, um die Modal-Positionierung zu steuern
ModalService.setShouldUseTopInsets = true;

type Props = {
  children: React.ReactNode;
  use_dark_theme?: boolean;
};
// Der Wrapper wird verwendet, um die Provider zu umschließen, die für die Komponenten notwendig sind
const AllTheProviders = ({ children, use_dark_theme = true }: Props) => {
  const theme = use_dark_theme
    ? { ...eva.dark, ...foofDarkTheme }
    : { ...eva.light, ...foofLightTheme };
  return (
    <>
      <IconRegistry icons={FontAwesomeIconsPack} />
      <ApplicationProvider {...eva} customMapping={mapping} theme={theme}>
        {children}
      </ApplicationProvider>
    </>
  );
};

// Erstellt einen Custom-Render, der die Provider um die Komponenten wickelt
const customRender = (
  component: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(component, { wrapper: AllTheProviders, ...options });
};

// Exportiert alle Funktionen der Testing Library, weswegen man diese dann auch von dieser Datei importieren kann
export * from "@testing-library/react-native";
export { customRender as render };
