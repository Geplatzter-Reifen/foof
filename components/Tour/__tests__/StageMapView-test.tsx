import React from "react";
import { Alert } from "react-native";
import { render } from "@/test-utils/test-utils";
import StageMapView from "../StagesMapView";
import { waitFor } from "@testing-library/react-native";

// Mock RNMapbox Maps
jest.mock("@rnmapbox/maps", () => ({
  MapView: ({ children }: { children?: React.ReactNode }) => (
    <div>{children}</div>
  ),
  Marker: () => <div>Mocked Marker</div>,
  Camera: () => <div>Mocked Camera</div>,
  locationManager: {
    start: jest.fn(),
    stop: jest.fn(),
  },
}));

// Mock Router and Navigation
const mockedNavigate = jest.fn();
const mockedGoBack = jest.fn();

jest.mock("expo-router", () => ({
  useLocalSearchParams: () => ({ tourId: "123" }),
  useRouter: () => ({
    back: mockedGoBack,
    navigate: mockedNavigate,
    push: jest.fn(),
  }),
  useNavigation: () => ({
    navigate: mockedNavigate,
    goBack: mockedGoBack,
    setOptions: jest.fn(),
  }),
}));

// Mock Alert
jest.spyOn(Alert, "alert");

describe("StageMapView Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("matches snapshot", async () => {
    const view = await waitFor(() =>
      render(<StageMapView tourId="test-tour-id" />),
    );
    expect(view.toJSON()).toMatchSnapshot();
  });
});
