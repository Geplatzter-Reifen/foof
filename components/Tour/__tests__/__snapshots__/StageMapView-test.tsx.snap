// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StageMapView Component matches snapshot 1`] = `
<div>
  <div>
    Mocked Camera
  </div>
</div>
`;
